1-development/ListPostsByRating/src/Api/Services/PostService.cs

```diff
@@ -0,0 +1,59 @@
+﻿using Api.Data;
+using Api.DTOs;
+using Api.Mappers;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Api.Services
+{
+    public interface IPostService
+    {
+        Task<PostDto> DownVotePostAsync(int id);
+        Task<IEnumerable<PostDto>> GetAllPostsAsync();
+        Task<PostDto> UpVotePostAsync(int id);
+    }
+
+    public class PostService : IPostService
+    {
+        private readonly IPostsRepository _postsRepository;
+
+        public PostService(IPostsRepository postsRepository)
+        {
+            _postsRepository = postsRepository ?? throw new System.ArgumentNullException(nameof(postsRepository));
+        }
+
+        public async Task<PostDto> DownVotePostAsync(int id)
+        {
+            // Find post
+            var postFound = await _postsRepository.GetByIdAsync(id)
+                .ConfigureAwait(false);
+            // Update DownVote
+            postFound.DownVotes++;
+            var updatedPost = await _postsRepository.UpdateAsync(postFound)
+                .ConfigureAwait(false);
+            return updatedPost.ToPostDto();
+        }
+
+        public async Task<IEnumerable<PostDto>> GetAllPostsAsync()
+        {
+            var results = await _postsRepository.GetAllAsync()
+                .ConfigureAwait(false);
+            var orderedPosts = results.OrderByDescending(post => post.Ratio)
+                .ThenByDescending(post => post.DiffVotes);
+            return orderedPosts.Select(post => post.ToPostDto());
+        }
+
+        public async Task<PostDto> UpVotePostAsync(int id)
+        {
+            // Find post
+            var postFound = await _postsRepository.GetByIdAsync(id)
+                .ConfigureAwait(false);
+            // Update UpVote
+            postFound.UpVotes++;
+            var updatedPost = await _postsRepository.UpdateAsync(postFound)
+                .ConfigureAwait(false);
+            return updatedPost.ToPostDto();
+        }
```

Comment:
Imagine multiple users are upvoting the same Post at the same time. Does this code guarantee we do not lose any upvotes? If the answer is 'yes', explain why. If the answer is 'no', tell us how it should be changed to prevent losing upvotes.

Reply:
The answer is 'no'. The model could probably be improved by using the [volatile](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/volatile) keyword for the member storing upvotes. The keyword assures the latest value available will be returned in concurrency situations.
Another workaround would be to set a lock before incrementing the upvotes count but I am not completely sure this would work in ASP .NET Core.
On the other hand, it can further be improved by using a database and choosing what kind of concurrency handling would be using, such as pessimistic concurrency or optimistic concurrency.
By using a pessimistic approach a database lock would be used to guarantee writes with the downside of performance issues.
By using an optimistic approach we set a field of type DateTime named `rowversion` in the model that stores the timestamp of the last write and in case of concurrency issues we would have to handle it with some kind of retry policy.